File: card/card.py
#-*-coding:utf-8-*-

def readFile(cmdList):
    with open("card.inp", "rb") as inp:
        getNum = inp.readline().split()
        for order in inp:
            cmdList.append(order.rstrip())
    return getNum

class cardDeck :
    def __init__(self, cardNum):
        self.card = [i for i in xrange(1,cardNum+1)]
        self.midIdx = 0

        if cardNum%2 == 0:
            self.midIdx = int(cardNum/2)
        else :
            self.midIdx = int(cardNum/2)+1

        self.top_mid = self.card[:self.midIdx]
        self.mid_bottom = self.card[self.midIdx:]

    def shuffle(self):
        self.card = self.top_mid
        shuffleIdx = 1
        for n in self.mid_bottom:
            self.card.insert(shuffleIdx, n)
            shuffleIdx += 2
        self.top_mid = self.card[:self.midIdx]
        self.mid_bottom = self.card[self.midIdx:]

    def cut(self, beg, end):
        cutPart = self.card[beg-1:end]
        del self.card[beg-1:end]
        self.card += cutPart
        self.top_mid = self.card[:self.midIdx]
        self.mid_bottom = self.card[self.midIdx:]

    def card_Top_Mid_Bottom(self):
        midIdx = int(len(self.card)/2)-1
        bottomIdx = len(self.card) - 1
        result = [str(self.card[0]), str(self.card[midIdx]), str(self.card[bottomIdx])]
        with open("card.out", "wb") as out:
            out.write(' '.join(result))

def main():
    cmd = []
    cardNum, operNum = map(int, readFile(cmd))
    # set cardNum, operNum and Command list

    pack = cardDeck(cardNum)
    for line in cmd: # read Command list by line
        if 'Shuffle' in line:
            pack.shuffle()
        if 'Cut' in line:
            cutCmd = line.split()
            cutCmd.pop(0) # remove 'Cut' command
            # cutCmd = ['cutBeginNum', 'cutEndNum']
            cutBeg, cutEnd = map(int, cutCmd)
            pack.cut(cutBeg, cutEnd)

    pack.card_Top_Mid_Bottom()
    return

if __name__ == '__main__':
    main()

File: card/card.cpp
#include <fstream>

#include <list>
#include <string>

#include <algorithm>
#include <iterator>

#define FOR(i,b,e) for((i)=(b); (i)<=(e); ++(i))

using namespace std;

using deck = list<int>; // 'deck' comes from deck of cards.

deck operator+(const deck& a, const deck& b)
{
    deck result(a.cbegin(), a.cend());
    result.insert(result.cend(), b.cbegin(), b.cend());
    return result;
}

class cardPack{
    deck card, toHalf, fromHalf;
    int cardNum=0, topCard=0, midCard=0, bottomCard=0, midPoint=0;
public:
    explicit cardPack(int n){
        cardNum = n;
        midPoint = (cardNum%2==0) ? cardNum/2 : cardNum/2+1;

        int i;
        FOR(i,1,cardNum/2)
            toHalf.push_back(i); // 1 2 3 4 5 6 7 8 ... cardNum/2
        FOR(i,cardNum/2+1,cardNum)
            fromHalf.push_back(i); // cardNum/2+1, cardNum/2+2, ..., cardNum
        card = toHalf+fromHalf;
    }
    void shuffle() {
        int cnt=0; // counting cards...

        if(!card.empty()) card.clear();

        card = move(toHalf); // move() function make "toHalf" list clear.
        auto shufflePos = card.begin();
        while(shufflePos!=card.end()) // cards in fromHalf deck goes to cards' intervals.
            card.splice(++shufflePos, fromHalf, fromHalf.begin());

        while(!fromHalf.empty()) // if "fromHalf" deck has more cards than "toHalf"deck
            card.splice(shufflePos, fromHalf, fromHalf.begin());
            // shufflePos == card.end()

        // ...Setting halves...
        for(auto it:card){
            if(cnt<midPoint)
                toHalf.push_back(it);
            else
                fromHalf.push_back(it);
            cnt++;
        }
        topCard = toHalf.front();
        midCard = *(next(card.begin(), cardNum/2-1));
        bottomCard = fromHalf.back();
    }

    void cut(int b, int e){
        toHalf.clear();
        fromHalf.clear();
        // clear previous decks to reset halves...

        int cnt = 0; // counting cards...

        auto cardPos = card.begin();
        auto begPos = next(cardPos, b-1);
        auto endPos = next(cardPos, e);
        // next(iter, int) function returns iterator come from advance(iter, int).

        card.splice(card.end(),card, begPos, endPos);

        // ...Setting halves...
        for(auto it:card){
            if(cnt<midPoint)
                toHalf.push_back(it);
            else
                fromHalf.push_back(it);
            cnt++;
        }
        topCard = toHalf.front();
        midCard = *(next(card.begin(), cardNum/2-1));
        bottomCard = fromHalf.back();
    }

    friend ofstream& operator<<(ofstream& ofs, const cardPack& c) {
        ofs << c.topCard << ' ' << c.midCard << ' ' << c.bottomCard;
        return ofs;
    }
};

int main()
{
    ifstream ifs("card.inp", ios::binary);
    ofstream ofs("card.out", ios::binary);
    int getNum, operNum, cutBeg, cutEnd;
    string cmd; // command

    ifs >> getNum >> operNum;

    cardPack pack(getNum);
    while(ifs >> cmd){
        if(cmd == "Shuffle"){
            pack.shuffle();
        }
        else if(cmd == "Cut"){
            ifs >> cutBeg >> cutEnd;
            pack.cut(cutBeg,cutEnd);
        }
    }
    ofs << pack;
    return 0;
}
